prime-router/src/main/kotlin/tokens/AuthenticatedClaims.kt

```diff
@@ -1,135 +1,227 @@
 package gov.cdc.prime.router.tokens
 
+import com.google.common.net.HttpHeaders
+import com.microsoft.azure.functions.HttpRequestMessage
 import gov.cdc.prime.router.Sender
-import io.jsonwebtoken.Claims
-import io.jsonwebtoken.Jwts
+import gov.cdc.prime.router.azure.HttpUtilities
+import gov.cdc.prime.router.azure.WorkflowEngine
+import gov.cdc.prime.router.common.Environment
+import org.apache.logging.log4j.kotlin.Logging
+
+const val DO_OKTA_AUTH = "okta"
+val authenticationFailure = HttpUtilities.errorJson("Authentication Failed")
+val authorizationFailure = HttpUtilities.errorJson("Unauthorized")
 
 /**
  * This represents a set of claims coming from Okta that have been authenticated (but may or may not
- * have been authorized). For convenience, certain claims have been pulled out into more readable/usable forms,
- * using helper methods in the class, called during object construction.
+ * have been authorized). For convenience, certain claims have been pulled out of the [jwtClaims]
+ * into more readable/usable forms, using helper methods in the class, called during object construction.
+ *
+ * The companion object has a number of helper methods that hide the Okta vs Server2Server differences.
  */
-class AuthenticatedClaims {
+class AuthenticatedClaims : Logging {
     /**
      * Raw claims map, typically as sent in a JWT
      */
     val jwtClaims: Map<String, Any>
 
-    // For Okta, these below are all derived from the raw jwtClaims.
+    // All of these below are _derived_ from the raw jwtClaims.
 
     /** Name of user as extracted from the subject claim.  Usually an email address */
     val userName: String
+
     /** Does this user have the prime administrator claim? */
     val isPrimeAdmin: Boolean
-    /** Name of the organization found in these claims */
-    val organizationNameClaim: String?
-    /** Is this a Sender claim, of the form DHSender_orgname  ? */
-    val isSenderOrgClaim: Boolean
+
+    /** Set of scopes associated with these claims.   Scopes are the "common language" between
+     * Okta and server2server auth.   Okta "DH" claims are mapped to scopes,
+     * or are created directly from server2server claims.
+     */
+    val scopes: Set<String>
 
     /**
-     * This constructor assumes certain claims that are found in Okta. Only use this for Okta claims.
+     * [_jwtClaims]  The raw list of claims
+     * [isOktaAuth] true if the claims came from Okta (eg, groups like "DHmd-phd").  false if the claims came from
+     * server2server auth (eg, scopes like md-phd.default.report)
+     *
+     * Each value passed must have been already authenticated.  This is for claims, *not* for required authorizations.
      */
-    constructor(_jwtClaims: Map<String, Any>) {
```

Comment:
In a future PR, we can get rid of this rather ugly`isOktaAuth` (and related `authentication-type: okta` header!), and simply check for whether the user sent `organization` Okta claims or `scope` Server2Server claims.  A bit of a hack, but worth it to avoid the annoying header.